File: spring-board/service/article/src/main/java/kopher/board/article/service/ArticleService.java
package kopher.board.article.service;

import kopher.board.article.entity.Article;
import kopher.board.article.repository.ArticleRepository;
import kopher.board.article.service.request.ArticleCreateRequest;
import kopher.board.article.service.request.ArticleUpdateRequest;
import kopher.board.article.service.response.ArticlePageResponse;
import kopher.board.article.service.response.ArticleResponse;
import kuke.board.common.snowflake.Snowflake;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ArticleService {

    private final Snowflake snowflake = new Snowflake();
    private final ArticleRepository articleRepository;

    @Transactional
    public ArticleResponse create(ArticleCreateRequest request) {
        Article article = articleRepository
                .save(Article.create(snowflake.nextId(), request.getTitle(), request.getContent(), request.getBoardId(), request.getWriterId()));
        return ArticleResponse.from(article);
    }

    @Transactional
    public ArticleResponse update(Long articleId, ArticleUpdateRequest request) {
        Article article = articleRepository.findById(articleId).orElseThrow();
        // 사실은 save 하는게 좋음
        article.update(request.getTitle(), request.getContent());
        return ArticleResponse.from(article);
    }

    public ArticleResponse read(Long articleId) {
        return ArticleResponse.from(articleRepository.findById(articleId).orElseThrow());
    }

    @Transactional
    public void delete(Long articleId) {
        articleRepository.deleteById(articleId);
    }

    public ArticlePageResponse readAll(Long boardId, Long page, Long pageSize) {
        return ArticlePageResponse.of(
                articleRepository.findAll(boardId, (page - 1) * pageSize, pageSize).stream()
                        .map(ArticleResponse::from)
                        .toList(),
                articleRepository.count(
                        boardId,
                        PageLimitCalculator.calculatePageLimit(page, pageSize, 10L)
                ));
    }
}

File: board/service/article/src/main/kotlin/jp/co/kopher/board/article/service/ArticleService.kt
package jp.co.kopher.board.article.service

import board.common.snowflake.Snowflake
import jp.co.kopher.board.article.entity.Article
import jp.co.kopher.board.article.repository.ArticleRepository
import jp.co.kopher.board.article.service.request.ArticleCreateRequest
import jp.co.kopher.board.article.service.request.ArticleUpdateRequest
import jp.co.kopher.board.article.service.response.ArticleResponse
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class ArticleService(
    private val articleRepository: ArticleRepository
) {
    private val snowflake: Snowflake = Snowflake()

    fun create(request: ArticleCreateRequest): ArticleResponse {
        val article = articleRepository.save(
            Article.create(snowflake.nextId(), request.title, request.content, request.boardId, request.writerId)
        )
        return ArticleResponse.from(article)
    }

    fun update(articleId: Long, request: ArticleUpdateRequest): ArticleResponse {
        val article = articleRepository.findByIdOrNull(articleId)
            ?: throw IllegalArgumentException("Article not found")
        article.update(request.title, request.content)
        return ArticleResponse.from(article)
    }

    @Transactional(readOnly = true)
    fun read(articleId: Long): ArticleResponse {
        return ArticleResponse.from(
            articleRepository.findByIdOrNull(articleId)
                ?: throw IllegalArgumentException("Article not found")
        )
    }

    fun delete(articleId: Long) {
        articleRepository.deleteById(articleId)
    }
}